### PR TITLE
ObjectOutliner: support outlining of classes in global let variables

### PR DESCRIPTION
**Pull request by eeckstein**
```
  let c = SomeClass()
```

is turned into

```
  private let outlinedVariable = SomeClass()  // statically initialized and allocated in the data section
  let c = outlinedVariable
```

rdar://111021230
rdar://115502043

This requires a change in SIL: add a dependency token operand to `global_addr`

Optionally, the dependency to the initialization of the global can be specified with a dependency token `depends_on <token>`.
This is usually a `builtin "once"` which calls the initializer for the global variable.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ObjectOutliner.swift [hunk 1]
```diff
@@ -314,23 +329,25 @@ private func constructObject(of allocRef: AllocRefInstBase,
...
+    // Create the initializers for the tail elements.
+    let numTailTupleElems = allocRef.numStoresPerTailElement
+    if numTailTupleElems > 1 {
+      // The elements are tuples: combine numTailTupleElems elements to a single tuple instruction.
+      for elementIdx in 0..<allocRef.numTailElements! {
+        var tupleElems = [Value]()
+        for tupleIdx in 0..<numTailTupleElems {
+          let store = storesToTailElements[elementIdx * numTailTupleElems + tupleIdx]
+          tupleElems.append(cloner.clone(store.source as! SingleValueInstruction))
+        }
```
Comment (karwa) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ObjectOutliner.swift:342]: Drive-by comment:

Since tupleElems has a known number of elements, it would be more efficient (and arguably clearer/easier to read) to use map:

```suggestion
        let tupleElems = (0..<numTailTupleElems).map { tupleIdx in
          let store = storesToTailElements[elementIdx * numTailTupleElems + tupleIdx]
          return cloner.clone(store.source as! SingleValueInstruction)
        }
```

Not a big deal but I just happened to notice it :)

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test
